### DISCUSSION TITLE
Is there a JCB way of doing nested sets? (storing items with hierarchical parent/child relationships)

### DISCUSSION THREAD

**Discussion by Obscerno** [joomengine/Joomla-Component-Builder#751] [Ideas]
Joomla has native support for nested sets, which is a method of storing hierarchical data. The bare minimum to getting them working is described here: https://docs.joomla.org/Using_nested_sets

The gist is they have made a table class you can inherit called [JTableNested](https://api.joomla.org/cms-3/classes/Joomla.CMS.Table.Nested.html) that does most of the work for you. You just have to add a few extra fields and add a bit of code to add the root node if it doesn't exist, make it the parent of items by default, and call functions to insert/reorder items.

JCB makes this pretty easy to implement with fields and custom code, using that article I linked to as reference. I've already done it in my project, I mainly just asked this as a signpost for future searchers, but if there is a way I've missed I'd love to know!

**Comment by Llewellynvdm** [comment]
Are you willing to share the code, with screenshots of the areas in JCB used... then I see if I can automate that for all ;) as this sounds like a very useful feature.

**Reply by Obscerno** [reply]
Yes I am! But I'll get back to you soon; I want to refactor a bit. :blush:

Also, list view sorting wasn't something I was planning to implement, but I feel like for JCB it would be best to include. I'll share the code that gets nested sets up and functioning, so it's here and at the very least might help someone else, and then I'll include list sorting down the road if I get some time!

**Reply by Obscerno** [reply]
@Llewellynvdm okay I've submitted an issue about this with the details: #752 

I don't include screenshots, but I can add those in later if you'd like! I don't have time right now.

**Reply by Obscerno** [reply]
Oh! And also, I did get around to ordering (I meant to say list view ordering up there). On the list view you can drag items and their nested orders are preserved. You can't set the order in the item though (the ordering value is rendered obsolete by this structure)
